nodefs/bridge.go

```diff
@@ -0,0 +1,553 @@
+	b.setEntryOutTimeout(out)
+	return fuse.OK
+}
+
+// addNewChild inserts the child into the tree. Returns file handle if file != nil.
+func (b *rawBridge) addNewChild(parent *Inode, name string, child *Inode, file File, out *fuse.EntryOut) uint64 {
+	lockNodes(parent, child)
+	parent.setEntry(name, child)
+	b.mu.Lock()
+	child.lookupCount++
+	if child.nodeID == 0 {
+		b.registerInode(child)
+	}
+
+	var fh uint64
+	if file != nil {
+		fh = b.registerFile(file)
+	}
+
+	out.NodeId = child.nodeID
+	// NOSUBMIT - or should let FS expose Attr.Ino? This makes
+	// testing semantics hard though, because os.Lstat doesn't
+	// reflect the FUSE FS
+	out.Attr.Ino = child.nodeID
+	out.Generation = b.nodes[out.NodeId].generation
+	b.mu.Unlock()
+	unlockNodes(parent, child)
+	return fh
+}
+
+func (b *rawBridge) setEntryOutTimeout(out *fuse.EntryOut) {
+	if b.options.AttrTimeout != nil {
+		out.SetAttrTimeout(*b.options.AttrTimeout)
+	}
+	if b.options.EntryTimeout != nil {
+		out.SetEntryTimeout(*b.options.EntryTimeout)
+	}
+}
+
+// registerInode sets an nodeID in the child. Must have bridge.mu and
+// child.mu
+func (b *rawBridge) registerInode(child *Inode) {
+	if l := len(b.free); l > 0 {
+		last := b.free[l-1]
+		b.free = b.free[:l-1]
+
+		child.nodeID = last
+		b.nodes[last].inode = child
+		b.nodes[last].generation++
+	} else {
+		last := len(b.nodes)
+		b.nodes = append(b.nodes, mapEntry{
+			inode: child,
+		})
+		child.nodeID = uint64(last)
+	}
+}
+
+func (b *rawBridge) Create(input *fuse.CreateIn, name string, out *fuse.CreateOut) (code fuse.Status) {
+	ctx := context.TODO()
+	parent, _ := b.inode(input.NodeId, 0)
+	child, f, flags, code := parent.node.Create(ctx, name, input.Flags, input.Mode)
+	if !code.Ok() {
+		if b.options.NegativeTimeout != nil {
+			out.SetEntryTimeout(*b.options.NegativeTimeout)
+		}
+		return code
+	}
+
+	out.Fh = b.addNewChild(parent, name, child, f, &out.EntryOut)
+	b.setEntryOutTimeout(&out.EntryOut)
+
+	out.OpenFlags = flags
+
+	var temp fuse.AttrOut
+	f.GetAttr(ctx, &temp)
+	out.Attr = temp.Attr
+	out.AttrValid = temp.AttrValid
+	out.AttrValidNsec = temp.AttrValidNsec
+
+	b.setEntryOutTimeout(&out.EntryOut)
+	if out.Attr.Mode&^07777 != fuse.S_IFREG {
+		log.Panicf("Create: mode must be S_IFREG (%o), got %o", fuse.S_IFREG, out.Attr.Mode)
+	}
+	return fuse.OK
+}
+
+func (b *rawBridge) Forget(nodeid, nlookup uint64) {
+	n, _ := b.inode(nodeid, 0)
+	n.removeRef(nlookup, false)
+}
+
+func (b *rawBridge) unregisterNode(nodeid uint64) {
+	b.free = append(b.free, nodeid)
+	b.nodes[nodeid].inode = nil
+}
+
+func (b *rawBridge) SetDebug(debug bool) {}
+
+func (b *rawBridge) GetAttr(input *fuse.GetAttrIn, out *fuse.AttrOut) fuse.Status {
+	n, fEntry := b.inode(input.NodeId, input.Fh())
+	f := fEntry.file
+
+	if input.Flags()&fuse.FUSE_GETATTR_FH == 0 {
+		f = nil
+	}
+
+	code = n.node.GetAttr(context.TODO(), f, out)
+	if out.Nlink == 0 {
+		// With Nlink == 0, newer kernels will refuse link
+		// operations.
+		out.Nlink = 1
```

Comment:
Can we get rid of this? Nlink = 0 can happen. It is checked by xfstests generic/035 and currently fails: https://github.com/hanwen/go-fuse/issues/137

Reply:
fixed now.